### PR TITLE
Dockerfile for scheduler

### PR DESCRIPTION
**Pull request by nicgrayson**
1. Provides Dockerfile for building the scheduler docker image
2. Hooks into existing build script
3. Added usage to Readme

Let me know if you have any questions.

### CODE REVIEW

**Comment by elingg** [comment]
Thanks @nicgrayson! I will take a look ASAP. My first comment is could you check in your marathon.json as a separate file?

**Comment by nicgrayson** [comment]
@elingg added and updated readme with example curl

File: marathon.json [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
+  "id": "hdfs-mesos",
+  "container": {
+    "type": "DOCKER",
+    "docker": {
+      "image": "mesosphere/hdfs-mesos",
```
Comment (BenWhitehead) [inline marathon.json]: We should come up with a scheme to use for tagging the docker images. Using `latest` will lead to issues down the road.

File: bin/build-hdfs [hunk 2]
```diff
@@ -91,3 +92,13 @@ cd ..
...
 rm -f $DIST.tgz
 tar czf $DIST.tgz $DIST
+
+
+# Build docker container
+if [ "$1" == "docker" ] && [ -f $DIST.tgz ]; then
+    if [ ! -z "$2" ]; then
+        DOCKER_IMAGE_NAME=$2
+    fi
+    docker build -t $DOCKER_IMAGE_NAME .
```
Comment (BenWhitehead) [inline bin/build-hdfs]: Update to allow specifying a tag when the docker image is being built.

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,13 @@
+FROM mesosphere/mesos:0.21.1-1.1.ubuntu1404
+
+ENV HDFS_MESOS_VERSION 0.0.2
```
Comment (BenWhitehead) [inline Dockerfile]: In docker it's a good idea to bind this env variable as late as possible, this is because running this directive will generate a new layer, which will invalidate any layer that was created after this directive.  In order to allow dockers layer caching to work as effectively as possible things should be bound as late as possible.

File: bin/build-hdfs [hunk 4]
```diff
@@ -9,6 +9,7 @@ IMPALA_ZIP=$IMPALA_VER.zip
 IMPALA_URL=https://github.com/cloudera/Impala/archive/$IMPALA_ZIP
 NATIVE=native
 DIST=hdfs-mesos-0.1.0
+DOCKER_IMAGE_NAME=hdfs-mesos
```
Comment (BenWhitehead) [inline bin/build-hdfs]: Change this to be `${DOCKER_IMAGE_NAME:-"hdfs-mesos"} so that the name can be overridden from outside the script if necessary.

File: bin/build-hdfs (hunk 2 above)
Comment (nicgrayson) [inline bin/build-hdfs]: If you pass in a tag it will run it now like when you run docker build:

`docker build -t repo/image:tag .`

File: bin/build-hdfs (hunk 4 above)
Comment (nicgrayson) [inline bin/build-hdfs]: It's the same variable as below. It's overridden if an image name is specified as $2

**Comment by nicgrayson** [comment]
@BenWhitehead I made some changes with your input. I think the default image name is pretty good now.

File: bin/build-hdfs (hunk 2 above)
Comment (BenWhitehead) [inline bin/build-hdfs]: ah, okay makes sense

File: bin/build-hdfs (hunk 4 above)
Comment (BenWhitehead) [inline bin/build-hdfs]: okay

**Comment by BenWhitehead** [comment]
With the latest changes I get an error when building the docker image:

**Comment by nicgrayson** [comment]
Fixed.

**Comment by BenWhitehead** [comment]
I can build cleanly now.

Not something that needs to be fixed now but in the future the may be the possibility to make the docker image more lean, right now it's effective size is ~700mb:
```
REPOSITORY                TAG                     IMAGE ID            CREATED              VIRTUAL SIZE
mesosphere/hdfs-mesos     0.1.0                   52d8604656df        About a minute ago   704.7 MB
```

**Comment by BenWhitehead** [comment]
@elingg should be okay to merge now.

**Comment by nicgrayson** [comment]
That is mostly due to how big the cdh binaries and therefore the dist
tarball is.

**Comment by BenWhitehead** [comment]
Fair enough, it's good that you've got a grasp on where the space is used.

**Comment by elingg** [comment]
Thanks @nicgrayson, @BenWhitehead . I took a look and it looks good, but I would like to host this on dockerhub and make building the docker file optional in the readme. If there are no objections, I will host the image on dockerhub and update the readme.

**Comment by nicgrayson** [comment]
@elingg The only problem with that is an automated build will not work on the the hub. This is because the dist targz is in the gitignore. That being said doing a build and push works. We should explain how to FROM that image to add your own config though.

**Comment by elingg** [comment]
@nicgrayson, yep our docker builders are not yet automated. That said, it would be nice to create an official docker image for our 0.1.0 release. The problem is that unless the user is using Mesos DNS, they will need to set `mesos.hdfs.zkfc.ha.zookeeper.quorum` in mesos-site.xml before building the docker file. Or they would have to run FROM the official docker image as you mentioned. Because of that, I think we can leave it as is for now. Let me know your final thoughts.

**Comment by rubbish** [comment]
It should be possible to override the hadoop configuration options with
java system properties. So it'd be nice to be able to pass those in via the
docker CMD. That would alleviate this problem.

**Comment by elingg** [comment]
Ok @rubbish, makes sense. @nicgrayson or @rubbish, would you mind updating the code, requiring the user to pass in this file as part of the docker CMD?

**Comment by elingg** [comment]
If so, I can build the docker file and publish it to the Mesosphere repo in dockerhub.

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+COPY hdfs-mesos-$HDFS_MESOS_VERSION.tgz /hdfs/
+RUN cd /hdfs && tar xfz hdfs-mesos-$HDFS_MESOS_VERSION.tgz
+
+ENV MESOS_NATIVE_JAVA_LIBRARY /usr/local/lib/libmesos.so
+ENV MESOS_NATIVE_LIBRARY /usr/local/lib/libmesos.so
+
+WORKDIR /hdfs/hdfs-mesos-$HDFS_MESOS_VERSION
+
+CMD ./bin/hdfs-mesos
```
Comment (rubbish) [inline Dockerfile]: Can we either modify this to be the entrypoint or change `/bin/hdfs-mesos` to accept `$JAVA_OPTS` to pass to java? That way we can override hadoop configuration with system properties.

Comment (elingg) [inline Dockerfile]: I feel like the simplest way is to pass the location of mesos-site.xml as an argument to the run command. If we do this through system properties, that sounds reasonable.

Comment (BenWhitehead) [inline Dockerfile]: Unless the mesos-site.xml is referenced as a URI that is downloaded into the sandbox there is now a container boundary that now has to be crossed. (can you shouldn't mount a file into the container using volumes. can cause lots of problems.)

Comment (elingg) [inline Dockerfile]: Assuming mesos-site.xml is in the current directory is a reasonable. We just need to document it.

Comment (nicgrayson) [inline Dockerfile]: After looking into this I determined that there is a fair amount of work in the ConfigServer to make it read from the system properties instead of the config file. Doing a FROM and building your own image to change the mesos-site.xml file seems like the best way for now.

**Comment by nicgrayson** [comment]
@elingg I think this PR is fine as is

File: Dockerfile (hunk 5 above)
Comment (rubbish) [inline Dockerfile]: I agree. If we were reading the config file into a org.apache.hadoop.conf.Configuration object and then rendering that, we would be able to override specific config values at the start of the framework with system properties i.e. passing `JAVA_OPTS="-Dmesos.master.uri=zk://somehost/mesos"` as an environmental variable to the start of the framework. This can be tackled in a separate PR.

**Comment by elingg** [comment]
@nicgrayson, okay, agreed I think this is fine for now. The problem is we can't upload it to dockerhub since there isn't a way to make mesos-site.xml configurable. If no final objections from anyone, I will merge. Thanks!

**Comment by BenWhitehead** [comment]
Another option would be to replace the start script with a custom script that would pre-process the config file based on environment variables, then launch the main process.

For an example of how this can easily be done using `envsubst` see this repo where I did it for kairosdb which is configured via properties file: https://github.com/mesosphere/docker-archlinux-kairosdb

**Comment by BenWhitehead** [comment]
Another node, using `FROM` suddenly creates a scenario where the docker image/tag will need to be configurable for all users since they would have to point to their version of the docker image.  This also complicates things further down the line when we want to release a new version of the framework. Now the user also has to rebuild their docker image based off the new image and causes a lot of friction.

**Comment by elingg** [comment]
@nicgrayson, @rubbish, what do you think of ben's idea above? I would prefer that so we could actually host the dockerfile and people could use it without building it themselves:

```Another option would be to replace the start script with a custom script that would pre-process the config file based on environment variables, then launch the main process.```

**Comment by rubbish** [comment]
Sounds good to me.

**Comment by nicgrayson** [comment]
@BenWhitehead @elingg This is ready to go with the env override changes

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+RUN cd /hdfs && tar xfz hdfs-mesos-$HDFS_MESOS_VERSION.tgz
+
+ENV MESOS_NATIVE_JAVA_LIBRARY /usr/local/lib/libmesos.so
+ENV MESOS_NATIVE_LIBRARY /usr/local/lib/libmesos.so
+
+ADD conf/mesos-site.xml /tmp/mesos-site.xml
+
+WORKDIR /hdfs/hdfs-mesos-$HDFS_MESOS_VERSION
+
+CMD ./bin/hdfs-mesos
```
Comment (BenWhitehead) [inline Dockerfile]: Please use the fully qualified path here to ensure that and potential change to workdir doesn't impact the ability to run the script.

**Comment by BenWhitehead** [comment]
One small change needed, then we can get a build setup for it and test it out.

File: Dockerfile (hunk 6 above)
Comment (nicgrayson) [inline Dockerfile]: @BenWhitehead done. This did require a change to the start script too.

**Comment by nicgrayson** [comment]
@BenWhitehead I reworked the run script so this could work. Should be good to go.

**Comment by BenWhitehead** [comment]
Thanks for the update @nicgrayson that looks okay to me.

Defer to @elingg for merging.

File: bin/hdfs-mesos [hunk 7]
```diff
@@ -1,5 +1,39 @@
 #!/bin/bash
 
+FRAMEWORK_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )/.."
+
+export MESOS_HDFS_DATA_DIR=${MESOS_HDFS_DATA_DIR:-"/tmp/hdfs/data"}
```
Comment (elingg) [inline bin/hdfs-mesos]: This looks great to me, but let's update the README to let users know they need to set the environment variables instead of directly modifying mesos-site.xml. It might work it to include a note mentioning that one important variable to configure is MESOS_HDFS_ZKFC_HA_ZOOKEEPER_QUORUM since if you use localhost here, the executors will point to localhost and not to the zk quorum.

Comment (nicgrayson) [inline bin/hdfs-mesos]: I pushed a readme update. You don't have to use env vars. Changing the mesos-site.xml directly works fine.

Comment (elingg) [inline bin/hdfs-mesos]: Ah, true. You can choose to overwrite the env variables and just change the mesos-site.xml directly. Thanks for updating the readme.

**Comment by elingg** [comment]
:+1:, I'm building and uploading the official docker image. Any final objections, @adam-mesos? If not, let's merge!

**Comment by elingg** [comment]
Also uploaded the docker file!

File: Dockerfile [hunk 8]
```diff
@@ -0,0 +1,17 @@
+FROM mesosphere/mesos:0.21.1-1.1.ubuntu1404
```
Comment (adam-mesos) [inline Dockerfile]: s/21.1/22.1/ ?

Comment (adam-mesos) [inline Dockerfile]: Or would 'latest' just work?

File: README.md [hunk 9]
```diff
@@ -55,8 +55,8 @@ Resource Reservation Instructions (Optional)
 5. On each slave with the new settings, stop the mesos slave by running
 <br>`sudo service mesos-slave stop`.</br>
 6. On each slave with the new settings, remove the old slave state by running
-<br>`rm -f /tmp/mesos/meta/slaves/latest`.</br> 
-Note: This will also remove task state, so you will want to manually kill any running tasks as a precaution.
+<br>`rm -f /tmp/mesos/meta/slaves/latest`</br>.
```
Comment (adam-mesos) [inline README.md]: Nit: Don't think we really need that trailing period '.'

File: README.md [hunk 10]
```diff
@@ -68,3 +68,23 @@ Shutdown Instructions (Optional)
 3. Access your zookeeper instance: `/PATH/TO/zookeeper/bin/zkCli.sh`
 4. Remove hdfs-mesos framework state from zookeeper: `rmr /hdfs-mesos`
 5. (Optional) Clear your data directories as specified in your `mesos-site.xml`. This is necessary to relaunch HDFS in the same directory.
+
+Docker Usage (Optional)
+--------------------------
+
+### Use [Official Mesosphere Image](https://registry.hub.docker.com/u/mesosphere/hdfs-mesos/)
```
Comment (adam-mesos) [inline README.md]: Can we add some basic info to the "official" docker image? Maybe a link to the DockerFile used to build it?

File: README.md [hunk 11]
```diff
@@ -68,3 +68,23 @@ Shutdown Instructions (Optional)
...
+- `./bin/build-hdfs docker`
+- If you wish to specify an image name run `./bin/build-hdfs docker <image-name>`
+
+### Run with Marathon
+
+Run the following command in the repository root to launch the docker container in your marathon install: `curl -X POST -H "Content-Type: application/json" http://<yourMarathonHost>:8080/v2/apps -d@marathon.json`
+
+### Configure [mesos-site.xml](conf/mesos-site.xml) with Environemnt Variables
+
+See [marathon.json](./marathon.json) for an example. Note you only need to set values you wish to override
```
Comment (adam-mesos) [inline README.md]: Nit: Add a trailing period '.' ?

File: bin/build-hdfs [hunk 12]
```diff
@@ -91,3 +93,15 @@ cd ..
 echo "Compressing to $DIST.tgz"
 rm -f $DIST.tgz
 tar czf $DIST.tgz $DIST
+
+
+# Build docker container
+if [ "$1" == "docker" ] && [ -f $DIST.tgz ]; then
+    if [ ! -z "$2" ]; then
```
Comment (adam-mesos) [inline bin/build-hdfs]: [ ! -z "foo" ] is equivalent to [ "foo" ], or you can use -n if you want to explicitly test for a non-empty string.

File: bin/hdfs-mesos [hunk 13]
```diff
@@ -1,5 +1,39 @@
...
+export MESOS_HDFS_JVM_OVERHEAD=${MESOS_HDFS_JVM_OVERHEAD:-1.35}
+export MESOS_HDFS_HADOOP_HEAP_SIZE=${MESOS_HDFS_HADOOP_HEAP_SIZE:-512}
+export MESOS_HDFS_NAMENODE_HEAP_SIZE=${MESOS_HDFS_NAMENODE_HEAP_SIZE:-4096}
+export MESOS_HDFS_DATANODE_HEAP_SIZE=${MESOS_HDFS_DATANODE_HEAP_SIZE:-1024}
+export MESOS_HDFS_EXECUTOR_HEAP_SIZE=${MESOS_HDFS_EXECUTOR_HEAP_SIZE:-256}
+export MESOS_HDFS_EXECUTOR_CPUS=${MESOS_HDFS_EXECUTOR_CPUS:-0.5}
+export MESOS_HDFS_NAMENODE_CPUS=${MESOS_HDFS_NAMENODE_CPUS:-1}
+export MESOS_HDFS_JOURNALNODE_CPUS=${MESOS_HDFS_JOURNALNODE_CPUS:-1}
+export MESOS_HDFS_DATANODE_CPUS=${MESOS_HDFS_DATANODE_CPUS:-1}
+export MESOS_HDFS_USER=${MESOS_HDFS_USER:-"root"}
```
Comment (adam-mesos) [inline bin/hdfs-mesos]: Would it make sense to just default this to empty string ("") so it inherits whatever user launched the hdfs-mesos scheduler?

File: bin/hdfs-mesos [hunk 14]
```diff
@@ -1,5 +1,39 @@
...
 
+FRAMEWORK_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )/.."
+
+export MESOS_HDFS_DATA_DIR=${MESOS_HDFS_DATA_DIR:-"/tmp/hdfs/data"}
+export MESOS_HDFS_SECONDARY_DATA_DIR=${MESOS_HDFS_SECONDARY_DATA_DIR:-"/var/run/hadoop-hdfs"}
+export MESOS_HDFS_NATIVE_HADOOP_BINARIES=${MESOS_HDFS_NATIVE_HADOOP_BINARIES:-false}
+export MESOS_HDFS_FRAMEWORK_MNT_PATH=${MESOS_HDFS_FRAMEWORK_MNT_PATH:-"/opt/mesosphere"}
+export MESOS_HDFS_STATE_ZK=${MESOS_HDFS_STATE_ZK:-"localhost:2181"}
+export MESOS_MASTER_URI=${MESOS_MASTER_URI:-"zk://localhost:2181/mesos"}
+export MESOS_HDFS_ZKFC_HA_ZOOKEEPER_QUORUM=${MESOS_HDFS_ZKFC_HA_ZOOKEEPER_QUORUM:-"localhost:2181"}
```
Comment (adam-mesos) [inline bin/hdfs-mesos]: These three definitely need to be overridden for a multi-node cluster. Let's make sure that's well documented. (Thanks if it already is!)

File: conf/mesos-site.xml [hunk 15]
```diff
@@ -20,131 +20,128 @@
...
 
   <property>
     <name>mesos.hdfs.journalnode.count</name>
     <description>Number of journal nodes (must be odd)</description>
-    <value>3</value>
+    <value>${MESOS_HDFS_JOURNALNODE_COUNT}</value>
   </property>
 
-  <!-- Additional settings for fine-tuning -->
-  <!--
```
Comment (adam-mesos) [inline conf/mesos-site.xml]: Now that you've uncommented all of these, there's no purpose to the defaults hardcoded in the source, right?

File: marathon.json [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+    "docker": {
+      "image": "mesosphere/hdfs-mesos:latest",
+      "network": "HOST"
+    }
+  },
+  "cpus": 0.1,
+  "mem": 256.0,
+  "instances": 1,
+  "env": {
+    "MESOS_HDFS_DATA_DIR": "/tmp/hdfs/data",
```
Comment (adam-mesos) [inline marathon.json]: Why should we need to manually set all these environment variables instead of falling back to the default values? I would think we would only need to specify any values that deviate from the defaults.

File: README.md (hunk 10 above)
Comment (adam-mesos) [inline README.md]: I just mean that when I go to https://registry.hub.docker.com/u/mesosphere/hdfs-mesos/ there is absolutely nothing under the "Information" tab. Users at least like to see a description of what the project is, and a link to the github repo hosting the Dockerfile, so they can know what they're pulling down, rather than randomly trusting that you're not installing malware/bloatware for them.
Would also be nice to get this into an automated build&publish setup, so it's clear that what's in Dockerhub is guaranteed to be the latest stable build of hdfs-mesos.

**Comment by adam-mesos** [comment]
Looks great! I'd just recommending upgrading the dependencies to Mesos 0.22.1 and adding info/Dockerfile to the dockerhub repo. Then :rocket:

File: Dockerfile (hunk 8 above)
Comment (BenWhitehead) [inline Dockerfile]: please don't depend on `latest` it is a moving target.

File: README.md (hunk 10 above)
Comment (nicgrayson) [inline README.md]: @elingg would need to update the docker hub page. This dockerfile will not work as an automated build currently.

File: bin/hdfs-mesos (hunk 13 above)
Comment (nicgrayson) [inline bin/hdfs-mesos]: I think we should leave that to another PR. I just copied in what was there before.

File: marathon.json (hunk 16 above)
Comment (nicgrayson) [inline marathon.json]: This file is intended to be an example. I just have them in here to show how it's done.

**Comment by nicgrayson** [comment]
All of @adam-mesos 's comments are resolved.

File: README.md (hunk 10 above)
Comment (BenWhitehead) [inline README.md]: Our teamcity hdfs build should be updated to push the docker image, and the repo should not be setup as an automated build repo on dockerhub.

Comment (adam-mesos) [inline README.md]: Any kind of automated build works for me.
As a user of this docker image, I would still like some Information about what I'm pulling/running.

File: bin/hdfs-mesos (hunk 13 above)
Comment (adam-mesos) [inline bin/hdfs-mesos]: ok

**Comment by adam-mesos** [comment]
LGTM. :rocket: !

File: README.md (hunk 10 above)
Comment (elingg) [inline README.md]: I will add information to dockerhub with a description and a link to the github repo. I will also create a JIRA to create an automated build to create the docker for each release.

**Comment by elingg** [comment]
Cool, I need to do the updates mentioned above related to docker and then will merge. Thanks for your patience, @nicgrayson!

File: README.md [hunk 17]
```diff
@@ -68,3 +68,31 @@ Shutdown Instructions (Optional)
...
+### Build (Optional)
+
+- `./bin/build-hdfs docker`
+- If you wish to specify an image name run `./bin/build-hdfs docker <image-name>`
+
+### Run with Marathon
+
+Run the following command in the repository root to launch the docker container in your marathon install: `curl -X POST -H "Content-Type: application/json" http://<yourMarathonHost>:8080/v2/apps -d@marathon.json`
+
+### Configure [mesos-site.xml](conf/mesos-site.xml) with Environemnt Variables
```
Comment (acmcelwee) [inline README.md]: Typo in `Environemnt` -> `Environment`

**Comment by F21** [comment]
What's the status on this one? I saw that it was reverted. Having it packaged as a docker container would make it much easier to deploy.

**Comment by elingg** [comment]
Hey @F21, this was indeed reverted. I really would like it to go in though after some more tweaks as the design is good. It just wasn't quite finished when it was merged.

**Comment by nicgrayson** [comment]
I want to note the reason this was reverted: https://github.com/mesosphere/hdfs/issues/133

**Comment by elingg** [comment]
Thanks for clarifying @nicgrayson
